scipy/linalg/basic.py

```diff
@@ -34,10 +40,13 @@ def _solve_check(n, info, lamch=None, rcond=None):
         return
     E = lamch('E')
     if rcond < E:
+        orig_formatwarning = warnings.formatwarning
+        warnings.formatwarning = _truncated_warning_msg
         warnings.warn('scipy.linalg.solve\nIll-conditioned matrix detected.'
                       ' Result is not guaranteed to be accurate.\nReciprocal '
-                      'condition number/precision: {} / {}'.format(rcond, E),
-                      RuntimeWarning)
+                      'condition number{:.6e}'.format(rcond),
+                      RuntimeWarning, stacklevel=1)
```

Comment:
stacklevel=3 I think

Reply:
If solve is called directly then stacklevel=3 will report -2th item on the tracestack instead of `basic.py`. I've highlighted the message below.

![image](https://user-images.githubusercontent.com/1303842/31246075-8c004034-aa0c-11e7-9ed9-d6dc7e44781d.png)

Reply:
Yes, and that's the correct place to emit the warning.
You can also find Numpy's np.deprecate decorator emits the warning with that stacklevel.

Reply:
Yes. Just to confirm if I understand correctly, this is only going to fix for linalg.solve warning and other functions which embed linalg.solve will be left as is (which was what Eric mentioned in the linked issue).

Reply:
Yes, I agree how to best deal with the warnings situation of nested calls to e.g. solve() is an open problem, given that python's warnings module is not fully up to the task imho.

Aside from solutions that require patching the warnings module (which I think we should not start doing), setting the stacklevel at least solves the issue in simplest cases, and we can hope the non-simple cases are not that pressing...

Another thing is that the emitted warning probably should be in a separate category to make these warnings easier to filter, `warnings.warn(..., category=scipy.linalg.LinAlgWarning)` (with `class LinAlgWarning(RuntimeWarning): ...` is added to scipy.linalg).